### ISSUE TITLE
Use Ripgrep as default search

### ISSUE THREAD

**Issue by st3soren** [redox-os/redox#746]
Could you possibly make Ripgrep the default CLI search tool for redox. [0]

It's also written in Rust.
It's much faster than grep, and other search utilities. [1]

[0] https://github.com/BurntSushi/ripgrep
[1] http://blog.burntsushi.net/ripgrep/

**Comment by jackpot51** [comment]
This is not possible right bow - it depends on a larfe number of crates and some do not support Redox

**Comment by stratact** [comment]
This issue should probably go into the coreutils repo?

**Comment by hauleth** [comment]
@st3soren no.

Simply because ripgrep and grep aren't compatible. ripgrep is similar-ish to `grep -R` which isn't expected as default behaviour from grep. However we could use parts of ripgrep to use as grep search engine.

**Comment by SShrike** [comment]
@hauleth Well, ripgrep uses the regex crate so I imagine you'd use that as the engine for grep?

**Comment by leeoniya** [comment]
maybe once https://github.com/BurntSushi/ripgrep/issues/162 lands, it should be possible to integrate ripgrep without `-R`

**Comment by BurntSushi** [comment]
Indeed, once I finish BurntSushi/ripgrep#162 (which realistically isn't *that* far away, it could conceivably be done by the end of this year), then it should be quite possible to write a fast `grep` that *mostly* respects its interface.

There are challenges though, which are strongly correlated with exactly how much compatibility you want to have with POSIX grep, for example. e.g., Standard `grep` uses something called *basic* regular expressions, which itself would require a special regex parser. (ripgrep operates more like `egrep` by default.) There are other problems, for example, grep supports backreferences and Rust's regex engine doesn't. There are probably also some subtle differences in how submatches/captures are matched too.

What I'm saying is that 100% grep compatibility is going to be hard, so if you really need that, you probably need to use GNU grep.

(With that said, I'm not particularly familiar with other challenges, like making ripgrep actually compile/work on Redox.)

**Comment by xTibor** [comment]
> This is not possible right now - it depends on a large number of crates and some do not support Redox

After a year there's only one crate left to port, [memmap-rs](https://github.com/danburkert/memmap-rs). With a dummy port ripgrep already builds and runs on Redox: https://i.imgur.com/rHcjZiE.png
